### ISSUE TITLE
ERROR: YASDI ini file was not found or is unreadable!

### ISSUE THREAD

**Issue by ShyamBhattacharyya** [konstantinblaesi/yasdi#1]
```
      $Bash>  cd <to the YASDI project path>
      $Bash>  cd projects/generic-cmake
      $Bash>  mkdir build-gcc
      $Bash>  cd build-gcc
      $Bash>  cmake ..
      $Bash>  make
      $Bash>  sudo make install  
```
After following these steps, when I am trying to run yasdishell from terminal or from projects/generic-cmake/build-gcc/yasdishell, the program runs but not reading the yasdi.ini file. Error Details: "ERROR: YASDI ini file was not found or is unreadable!". I placed yasdi.ini file both in projects/generic-cmake/build-gcc/ and /usr/local/lib folder. I have also tried to compile the sample.c code and run. but still same issue. **Please guide me where to place the yasdi.ini file after the sudo make install command.**

**Comment by ShyamBhattacharyya** [comment]
This is my directory structure.
```
~/YASDI/projects/generic-cmake/build-gcc
.
├── CMakeCache.txt
├── CMakeFiles
│   ├── 3.16.3
│   │   ├── CMakeCCompiler.cmake
│   │   ├── CMakeDetermineCompilerABI_C.bin
│   │   ├── CMakeSystem.cmake
│   │   └── CompilerIdC
│   │       ├── a.out
│   │       ├── CMakeCCompilerId.c
│   │       └── tmp
│   ├── cmake.check_cache
│   ├── CMakeDirectoryInformation.cmake
│   ├── CMakeOutput.log
│   ├── CMakeTmp
│   ├── Makefile2
│   ├── Makefile.cmake
│   ├── progress.marks
│   ├── TargetDirectories.txt
│   ├── yasdi.dir
│   │   ├── build.make
│   │   ├── C.includecache
│   │   ├── cmake_clean.cmake
│   │   ├── DependInfo.cmake
│   │   ├── depend.internal
│   │   ├── depend.make
│   │   ├── flags.make
│   │   ├── home
│   │   │   └── pi
│   │   │       └── YASDI
│   │   │           ├── core
│   │   │           │   ├── byteorder.o
│   │   │           │   ├── defractionizer.o
│   │   │           │   ├── driver_layer.o
│   │   │           │   ├── fractionizer.o
│   │   │           │   ├── iorequest.o
│   │   │           │   ├── mempool.o
│   │   │           │   ├── minmap.o
│   │   │           │   ├── minqueue.o
│   │   │           │   ├── netpacket.o
│   │   │           │   ├── prot_layer.o
│   │   │           │   ├── repository.o
│   │   │           │   ├── router.o
│   │   │           │   ├── scheduler.o
│   │   │           │   ├── smadata_layer.o
│   │   │           │   ├── statistic_writer.o
│   │   │           │   ├── timer.o
│   │   │           │   └── tools.o
│   │   │           ├── libs
│   │   │           │   └── libyasdi.o
│   │   │           ├── os
│   │   │           │   └── os_linux.o
│   │   │           ├── protocol
│   │   │           │   ├── smanet.o
│   │   │           │   └── sunnynet.o
│   │   │           └── smalib
│   │   │               └── getini.o
│   │   ├── link.txt
│   │   └── progress.make
│   ├── yasdi_drv_ip.dir
│   │   ├── build.make
│   │   ├── C.includecache
│   │   ├── cmake_clean.cmake
│   │   ├── DependInfo.cmake
│   │   ├── depend.internal
│   │   ├── depend.make
│   │   ├── flags.make
│   │   ├── home
│   │   │   └── pi
│   │   │       └── YASDI
│   │   │           └── driver
│   │   │               └── ip_generic.o
│   │   ├── link.txt
│   │   └── progress.make
│   ├── yasdi_drv_serial.dir
│   │   ├── build.make
│   │   ├── C.includecache
│   │   ├── cmake_clean.cmake
│   │   ├── DependInfo.cmake
│   │   ├── depend.internal
│   │   ├── depend.make
│   │   ├── flags.make
│   │   ├── home
│   │   │   └── pi
│   │   │       └── YASDI
│   │   │           └── driver
│   │   │               └── serial_posix.o
│   │   ├── link.txt
│   │   └── progress.make
│   ├── yasdimaster.dir
│   │   ├── build.make
│   │   ├── C.includecache
│   │   ├── cmake_clean.cmake
│   │   ├── DependInfo.cmake
│   │   ├── depend.internal
│   │   ├── depend.make
│   │   ├── flags.make
│   │   ├── home
│   │   │   └── pi
│   │   │       └── YASDI
│   │   │           ├── libs
│   │   │           │   └── libyasdimaster.o
│   │   │           └── master
│   │   │               ├── busevents.o
│   │   │               ├── chanvalrepo.o
│   │   │               ├── main.o
│   │   │               ├── mastercmd.o
│   │   │               ├── netchannel.o
│   │   │               ├── netdevice.o
│   │   │               ├── objman.o
│   │   │               ├── plant.o
│   │   │               ├── stateconfig.o
│   │   │               ├── statedetection.o
│   │   │               ├── stateident.o
│   │   │               ├── statereadchan.o
│   │   │               ├── statewritechan.o
│   │   │               └── ysecurity.o
│   │   ├── link.txt
│   │   └── progress.make
│   └── yasdishell.dir
│       ├── build.make
│       ├── C.includecache
│       ├── cmake_clean.cmake
│       ├── DependInfo.cmake
│       ├── depend.internal
│       ├── depend.make
│       ├── flags.make
│       ├── home
│       │   └── pi
│       │       └── YASDI
│       │           └── shell
│       │               └── CommonShellUIMain.o
│       ├── link.txt
│       └── progress.make
├── cmake_install.cmake
├── CTestTestfile.cmake
├── devices
├── install_manifest.txt
├── libyasdi_drv_ip.so -> libyasdi_drv_ip.so.1
├── libyasdi_drv_ip.so.1 -> libyasdi_drv_ip.so.1.8.1
├── libyasdi_drv_ip.so.1.8.1
├── libyasdi_drv_serial.so -> libyasdi_drv_serial.so.1
├── libyasdi_drv_serial.so.1 -> libyasdi_drv_serial.so.1.8.1
├── libyasdi_drv_serial.so.1.8.1
├── libyasdimaster.so -> libyasdimaster.so.1
├── libyasdimaster.so.1 -> libyasdimaster.so.1.8.1
├── libyasdimaster.so.1.8.1
├── libyasdi.so -> libyasdi.so.1
├── libyasdi.so.1 -> libyasdi.so.1.8.1
├── libyasdi.so.1.8.1
├── Makefile
├── yasdi.ini
└── yasdishell
```

**Comment by konstantinblaesi** [comment]
@ShyamBhattacharyya This repository is mostly a mirror of the code published by SMA and unfortunately the code base is not very nice. Sometimes it leads to confusing output :) The message 'ERROR: YASDI ini file was not found or is unreadable!' is also printed when the ini file was readable but the library failed to initialize. I noticed two things:

1. by default it looks for yasdi.ini in the yasdishell directory, but you can also pass a path to a different location using for example `./yasdishell /path/to/yasdi.ini`
2. when running it from the build directory (so without 'make install' to install it to the system paths) it doesn't find the libraries in the build directory and the library fails to initialize because there are no drivers (serial or ip)

Running from the build directory like this works (sma.ini from my setup)

```
pi@pi3:~/yasdi/projects/generic-cmake/build $ cat yasdi.ini 
[DriverModules]
Driver0=libyasdi_drv_serial

[COM1]
Device=/dev/ttyUSB0
Media=RS485
Baudrate=1200
Protocol=SMANet

[Misc]
DebugOutput=/dev/stderr

pi@pi3:~/yasdi/projects/generic-cmake/build $ ll /dev/ttyUSB0 
crw-rw---- 1 root dialout 188, 0 23. Apr 21:55 /dev/ttyUSB0

pi@pi3:~/yasdi/projects/generic-cmake/build $ ll
total 396K
-rw-r--r-- 1 pi   pi    14K 23. Apr 17:28 CMakeCache.txt
drwxr-xr-x 9 pi   pi   4,0K 23. Apr 17:42 CMakeFiles
-rw-r--r-- 1 pi   pi    11K 23. Apr 17:28 cmake_install.cmake
-rw-r--r-- 1 pi   pi    521 23. Apr 17:28 CTestTestfile.cmake
drwxr-xr-x 2 pi   pi   4,0K 23. Apr 17:29 devices
-rw-r--r-- 1 root root  453 23. Apr 17:42 install_manifest.txt
lrwxrwxrwx 1 pi   pi     20 23. Apr 17:29 libyasdi_drv_ip.so -> libyasdi_drv_ip.so.1
lrwxrwxrwx 1 pi   pi     24 23. Apr 17:29 libyasdi_drv_ip.so.1 -> libyasdi_drv_ip.so.1.8.1
-rwxr-xr-x 1 pi   pi    19K 23. Apr 17:29 libyasdi_drv_ip.so.1.8.1
lrwxrwxrwx 1 pi   pi     24 23. Apr 17:29 libyasdi_drv_serial.so -> libyasdi_drv_serial.so.1
lrwxrwxrwx 1 pi   pi     28 23. Apr 17:29 libyasdi_drv_serial.so.1 -> libyasdi_drv_serial.so.1.8.1
-rwxr-xr-x 1 pi   pi    19K 23. Apr 17:29 libyasdi_drv_serial.so.1.8.1
lrwxrwxrwx 1 pi   pi     19 23. Apr 17:29 libyasdimaster.so -> libyasdimaster.so.1
lrwxrwxrwx 1 pi   pi     23 23. Apr 17:29 libyasdimaster.so.1 -> libyasdimaster.so.1.8.1
-rwxr-xr-x 1 pi   pi   116K 23. Apr 17:29 libyasdimaster.so.1.8.1
lrwxrwxrwx 1 pi   pi     13 23. Apr 17:29 libyasdi.so -> libyasdi.so.1
lrwxrwxrwx 1 pi   pi     17 23. Apr 17:29 libyasdi.so.1 -> libyasdi.so.1.8.1
-rwxr-xr-x 1 pi   pi   124K 23. Apr 17:29 libyasdi.so.1.8.1
-rw-r--r-- 1 pi   pi    41K 23. Apr 17:28 Makefile
-rwxrwxrwx 1 pi   pi    147 23. Apr 22:25 yasdi.ini
-rwxr-xr-x 1 pi   pi    24K 23. Apr 17:29 yasdishell
pi@pi3:~/yasdi/projects/generic-cmake/build $ cat yasdi.ini 
[DriverModules]
Driver0=libyasdi_drv_serial

[COM1]
Device=/dev/ttyUSB0
Media=RS485
Baudrate=1200
Protocol=SMANet

[Misc]
DebugOutput=/dev/stderr

pi@pi3:~/yasdi/projects/generic-cmake/build $ LD_LIBRARY_PATH=$(pwd) ./yasdishell 
************************************************************

        YASDI Mini Shell (build for Linux) 

      (This is an test program using YASDI)

************************************************************
Switching driver 'COM1' on...success

Command ('?' for help):
```

**Comment by konstantinblaesi** [comment]
@ShyamBhattacharyya in case you want to install it system wide, see https://github.com/konstantinblaesi/sma-yasdi-logger/blob/master/docker/Dockerfile#L38 for an example

**Comment by konstantinblaesi** [comment]
@ShyamBhattacharyya I have restructured the repository a bit and made container images for amd64 and arm64 available (alpine and debian based), see https://github.com/konstantinblaesi/yasdi/pkgs/container/yasdi/versions for details

**Comment by ShyamBhattacharyya** [comment]
Thank you @konstantinblaesi for your prompt support. I have also successfully built my requirement using the API written in sample.c code given in the repository. Some errors are present in the function name inside the sample.c. It will not compile. First, I was  very confused about the terms like device handle, channel handle, channel name etc in contrast with Modbus protocol. The pdf document instruction (what I found on the internet) is not that good to write your own code. At that time Dockerfile was also not available in the last repository. Now the updated repo is good for a newbie. Another thing I want to add on top of that is, if you provide a Dockerfile for system arch like arml6v and arm7 for RPI(zero and rpi-3,4 etc) usage, it will be a bang!! Thank you for you good work and maintaining the repo.
